File: 7Modules/Melrick/frontend/src/types/analytics.ts
export type KPIOverview = {
  total_orders: number;
  open_orders: number;
  delivered_orders: number;
  cancelled_orders: number;
  active_deliveries: number;
  total_revenue_inr: number;
  avg_match_score: number;
  avg_supplier_reliability: number;
  low_stock_items: number;
  optimization_savings_km: number;
  optimization_savings_percent: number;
};

export type StatusCount = {
  status: string;
  count: number;
};

export type MatchingAnalytics = {
  total_items_scored: number;
  avg_candidates_per_item: number;
  avg_top_score: number;
  avg_selected_score: number;
  urgency_top_score: Record<string, number>;
};

export type DeliveryAnalytics = {
  total_deliveries: number;
  planned_deliveries: number;
  in_progress_deliveries: number;
  completed_deliveries: number;
  avg_distance_km: number;
  avg_duration_minutes: number;
  total_naive_distance_km: number;
  total_optimized_distance_km: number;
  total_savings_km: number;
  total_savings_percent: number;
};

export type SupplierPerformance = {
  supplier_id: number;
  supplier_name: string;
  reliability_score: number;
  assignments_total: number;
  fulfilled_assignments: number;
  rejected_assignments: number;
  fulfillment_rate: number;
  avg_match_score: number;
};

export type InventoryRiskItem = {
  catalog_id: number;
  supplier_id: number;
  supplier_name: string;
  part_number: string;
  part_name: string;
  current_stock: number;
  reorder_point: number;
};

export type EventTimelinePoint = {
  day: string;
  event_type: string;
  count: number;
};

export type AnalyticsSnapshot = {
  overview: KPIOverview;
  orders: StatusCount[];
  matching: MatchingAnalytics;
  deliveries: DeliveryAnalytics;
  suppliers: SupplierPerformance[];
  low_stock: InventoryRiskItem[];
  events: EventTimelinePoint[];
};

File: 7Modules/Melrick/backend/schemas/analytics.py
from __future__ import annotations

from datetime import date
from typing import Dict, List

from pydantic import BaseModel, Field


class KPIOverview(BaseModel):
    total_orders: int
    open_orders: int
    delivered_orders: int
    cancelled_orders: int
    active_deliveries: int
    total_revenue_inr: float
    avg_match_score: float
    avg_supplier_reliability: float
    low_stock_items: int
    optimization_savings_km: float
    optimization_savings_percent: float


class StatusCount(BaseModel):
    status: str
    count: int


class MatchingAnalytics(BaseModel):
    total_items_scored: int
    avg_candidates_per_item: float
    avg_top_score: float
    avg_selected_score: float
    urgency_top_score: Dict[str, float] = Field(default_factory=dict)


class DeliveryAnalytics(BaseModel):
    total_deliveries: int
    planned_deliveries: int
    in_progress_deliveries: int
    completed_deliveries: int
    avg_distance_km: float
    avg_duration_minutes: float
    total_naive_distance_km: float
    total_optimized_distance_km: float
    total_savings_km: float
    total_savings_percent: float


class SupplierPerformance(BaseModel):
    supplier_id: int
    supplier_name: str
    reliability_score: float
    assignments_total: int
    fulfilled_assignments: int
    rejected_assignments: int
    fulfillment_rate: float
    avg_match_score: float


class InventoryRiskItem(BaseModel):
    catalog_id: int
    supplier_id: int
    supplier_name: str
    part_number: str
    part_name: str
    current_stock: int
    reorder_point: int


class EventTimelinePoint(BaseModel):
    day: date
    event_type: str
    count: int


class AnalyticsSnapshot(BaseModel):
    overview: KPIOverview
    orders: List[StatusCount]
    matching: MatchingAnalytics
    deliveries: DeliveryAnalytics
    suppliers: List[SupplierPerformance]
    low_stock: List[InventoryRiskItem]
    events: List[EventTimelinePoint]
